agent/src/util/discovery_operator.rs

```diff
@@ -450,6 +475,223 @@ impl DiscoveryOperator {
+                    self.config.metadata.namespace.as_ref().unwrap(),
+                    secret_key_selector,
+                )
+                .await
+            }
+        }
+    }
+}
+
+/// This provides a mockable way to query configMap and secret
+#[cfg_attr(test, automock)]
+#[tonic::async_trait]
+pub trait KubeClient: Send + Sync {
+    async fn get_secret(&self, name: &str, namespace: &str) -> anyhow::Result<Option<Secret>>;
+
+    async fn get_config_map(
+        &self,
+        name: &str,
+        namespace: &str,
+    ) -> anyhow::Result<Option<ConfigMap>>;
+}
+
+struct ActualKubeClient {
+    pub kube_interface: Arc<dyn k8s::KubeInterface>,
+}
+
+impl ActualKubeClient {
+    pub fn new(kube_interface: Arc<dyn k8s::KubeInterface>) -> Self {
+        ActualKubeClient { kube_interface }
+    }
+}
+
+#[tonic::async_trait]
+impl KubeClient for ActualKubeClient {
+    async fn get_secret(&self, name: &str, namespace: &str) -> anyhow::Result<Option<Secret>> {
+        let resource_client: Api<Secret> =
+            Api::namespaced(self.kube_interface.get_kube_client(), namespace);
+        let resource = resource_client.get_opt(name).await?;
+        Ok(resource)
+    }
+
+    async fn get_config_map(
+        &self,
+        name: &str,
+        namespace: &str,
+    ) -> anyhow::Result<Option<ConfigMap>> {
+        let resource_client: Api<ConfigMap> =
+            Api::namespaced(self.kube_interface.get_kube_client(), namespace);
+        let resource = resource_client.get_opt(name).await?;
+        Ok(resource)
+    }
+}
+
+async fn get_discovery_property_value_from_secret(
+    kube_client: &dyn KubeClient,
+    namespace: &str,
+    secret_key_selector: &SecretKeySelector,
+) -> anyhow::Result<Option<ByteData>> {
+    let optional = secret_key_selector.optional.unwrap_or_default();
+    if secret_key_selector.name.is_none() {
+        if optional {
+            return Ok(None);
+        }
+        return Err(Error::new(ErrorKind::InvalidInput, "secret name is none").into());
+    }
+    let secret_key = &secret_key_selector.key;
+    let secret_name = secret_key_selector.name.as_ref().unwrap();
+
+    let secret = kube_client.get_secret(secret_name, namespace).await?;
+    if secret.is_none() {
+        if optional {
+            return Ok(None);
+        } else {
+            return Err(Error::new(ErrorKind::InvalidInput, "secret not found").into());
+        }
+    }
+    let secret = secret.unwrap();
+    // All key-value pairs in the stringData field are internally merged into the data field
+    // we don't need to check string_data.
+    if let Some(data) = secret.data {
+        if let Some(v) = data.get(secret_key) {
+            return Ok(Some(ByteData {
+                vec: Some(v.0.clone()),
+            }));
+        }
+    }
+
+    // secret key/value not found
+    if optional {
+        Ok(None)
+    } else {
+        Err(Error::new(ErrorKind::InvalidInput, "secret data not found").into())
```

Comment:
nit: might be helpful to have a tiny bit more context in error messages.  something like "discovery property's referenced secret ..."  (same for ConfigMap code below ... also, might use k8s capitalization rather than rust ... i.e. ConfigMap rather than config_map)

Reply:
add more context to the error messages